zip-tor-txn-relay-network.rst

```diff
@@ -0,0 +1,132 @@
+::
+
+  ZIP: ??? <TTRN>
+  Title: The Zcash Tor Transaction Relay Network
+  Owners: Nathan Wilcox <nathan@electriccoin.co>
+  Credits: 
+  Status: Proposed
+  Category: Application Standard
+  Created: 2020-08-12
+  License: MIT
+
+
+Terminology
+===========
+
+The key words "MUST", "SHOULD", and "OPTIONAL" in this document are to be interpreted
+as described in RFC 2119. [#RFC2119]_
+
+The term "network upgrade" in this document is to be interpreted as described in ZIP 200
+[#zip-0200]_.
+
+The term "Sapling" in this document is to be interpreted as described in ZIP 205
+[#zip-0205]_.
+
+The term "Sprout value pool balance" in this document is to be interpreted as described
+in ZIP 209 [#zip-0209]_.
+
+
+Abstract
+========
+
+The Zcash `Tor Transaction Relay Network` (`TTRN`) is an auxillary Zcash protocol to improve network privacy of wallets which issue transactions. TTRN `Service Providers` publish a `Service Descriptor` to a well-known Wishing Well (see [FIXME link to zip-wishing-well]), which includes a `registration fee` paid into the well (ie burned). TTRN-enabled wallets subscribe to that Wishing Well to discover Providers. When sending a shielded transaction the wallet selects a random provider, connects to it over Tor, and delivers the transaction via a TTRN-specific `Delivery Protocol`. These transactions pay `TTRN service fees` to providers.
+
+As an auxillary protocol, it is transparent and backwards compatible with the Zcash P2P protocol and consensus layer, and it is also opt-in for wallet usage. Light wallets, as well as full-node wallets, may be able to protect an important aspect of their privacy from hosted service providers with this protocol.
+
+Motivation
+==========
+
+When wallets generate transactions they must transmit them to the Zcash network, and this initial transmission is a weak point for Zcash network privacy: it is possible for network peers to determine which node originated a transaction, even when the P2P protocol is routed over Tor [#]. By associating two different fully shielded transactions with the same originating peer, those transactions are linked by this network metadata, thereby violating a key privacy property that Zeroknowledge proofs protect against on-chain analysis. When a peer does not route its P2P traffic over Tor, this association can also include an IP address, which can reveal some information about the physical location of the wallet.
```

Comment:
My understanding of the weakness is that p2p connections are long lived, so the topology is relatively stable. By using the simple heuristic that "if I see txn T from my neighbor A before my neighbor B, then the sending node must live somewhere in A's direction". With enough nodes, or enough repeated sends from a known target, it may be possible over time to narrow down a sender to a specific location in the topology.

This is just hand-wavy description I've heard second or third hand, so it would be useful for any privacy researchers who've looked at this exact problem to chime in. I would assume this same issue is in the scope of the Dandelion threat model, so I would look there first. (I'm pretty time constrained so I may not get to that research survey anytime soon. :-/ )

Reply:
@teor2345 the details are in this paper.  Skip to section 3C for information about Tor. https://arxiv.org/abs/1410.6079

1. attacker can run exit nodes that only connect to target peers.
2. attacker can use a bitcoin DoS protection mechanism to ban Tor exit nodes not controlled by attacker. (Not sure if this is still possible in Bitcoin or Zcash.)
3. attacker can DoS any Tor hidden services they don't control.
4. it's even cheaper to run sybil attacks with Tor hidden services than with IP addresses, since they're cheaper to create than IP addresses are to obtain.
5. attacker can probably become all of a Tor user's peers, know what transactions they are creating, and also manipulate their view of the network.